substrate/emissions/pallet/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+      let rewards_per_network = if pre_ec_security {
+        distances
+          .into_iter()
+          .map(|(n, distance)| {
+            // calculate how much each network gets based on distance to ec-security
+            let reward = u64::try_from(
+              u128::from(reward_this_epoch).saturating_mul(u128::from(distance)) /
+                u128::from(total_distance),
+            )
+            .unwrap();
+            (n, reward)
+          })
+          .collect::<BTreeMap<NetworkId, u64>>()
+      } else {
+        volume_per_network
+          .into_iter()
+          .map(|(n, vol)| {
+            // 20% of the reward goes to the Serai network and rest is distributed among others
+            // based on swap-volume.
+            let reward = if n == NetworkId::Serai {
+              reward_this_epoch / 5
+            } else {
+              let reward = reward_this_epoch - (reward_this_epoch / 5);
+              // TODO: It is highly unlikely but what to do in case of 0 total volume?
+              if total_volume != 0 {
+                u64::try_from(
+                  u128::from(reward).saturating_mul(u128::from(vol)) / u128::from(total_volume),
+                )
+                .unwrap()
+              } else {
+                0
+              }
+            };
+            (n, reward)
+          })
+          .collect::<BTreeMap<NetworkId, u64>>()
+      };
+
+      // distribute the rewards within the network
+      for (n, reward) in rewards_per_network {
+        let (validators_reward, pool_reward) = if n == NetworkId::Serai {
+          (reward, 0)
+        } else {
+          // calculate pool vs validator share
+          let capacity = ValidatorSets::<T>::total_allocated_stake(n).unwrap_or(Amount(0)).0;
+          let required = ValidatorSets::<T>::required_stake_for_network(n);
+          let unused_capacity = capacity.saturating_sub(required);
+
+          let distribution = unused_capacity.saturating_mul(ACCURACY_MULTIPLIER) / capacity;
+          let total = DESIRED_DISTRIBUTION.saturating_add(distribution);
+
+          let validators_reward = DESIRED_DISTRIBUTION.saturating_mul(reward) / total;
+          let pool_reward = reward.saturating_sub(validators_reward);
+          (validators_reward, pool_reward)
+        };
+
+        // distribute validators rewards
+        if Self::distribute_to_validators(n, validators_reward).is_err() {
+          // TODO: log the failure
+          continue;
+        }
+
+        // send the rest to the pool
+        let coin_count = u64::try_from(n.coins().len()).unwrap();
+        for c in n.coins() {
+          // assumes reward is equally distributed between network coins.
+          if Coins::<T>::mint(
+            Dex::<T>::get_pool_account(*c),
+            Balance { coin: Coin::Serai, amount: Amount(pool_reward / coin_count) },
+          )
+          .is_err()
+          {
+            // TODO: log the failure
```

Comment:
Can minting SRI fail? I think we can panic here without too much concern (though sure, a panic at the start of the block isn't great).

Reply:
there is `is_allowed` check being done on the SRI as well, so either that or overflow. I just think any panic in the hooks will stop the network since everyone will panic? or substrate just retries a new block?

Reply:
I'm unsure. I'd assume it fail the block, and move to the next block, which would stall the network *if all blocks at that position have some error*. This is something we should check.

Heard on overflow though. I still don't like having this error condition. This should be an invariant and any attempted handling of this invariant will always be incomplete to some degree. Solely logging there's an invariant doesn't change the presence of the invariant.

Reply:
Yea but the idea is that this is a rewards distribution code and we know it shouldn't fail. If it still fails for a node, just that node won't get the rewards for that session and that is all there is. We can see it in the logs and decide whether it is the result of a bug or just something normal and if it turns out to be a bug we can fix it. Which is better than panicking IMO. Because if we panic for the first node in the array we won't send any rewards to any node, if we log the error and continue we at least have chance at rewarding other nodes.  If you check the latest version of the code that I pushed yesterday this should be more obvious.